### PR TITLE
Add utility for CkArrayIndex, ArrayComponentId, Tensor, VolumeData and DataBoxTag

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

A number of small utility functions that are useful for #1527

### CODE REVIEW

File: src/Parallel/ArrayIndex.hpp [hunk 1]
```diff
@@ -44,13 +49,25 @@ struct ArrayIndex : public CkArrayIndex {
     nInts = sizeof(array_index) / sizeof(int);  // NOLINT
   }
 
+  // clang-tidy: mark explicit: it's a conversion constructor
+  ArrayIndex(const CkArrayIndex& array_index)  // NOLINT
+      : CkArrayIndex(array_index),
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        array_index_(reinterpret_cast<Index*>(CkArrayIndex::data())) {
+    ASSERT(CkArrayIndex::nInts == sizeof(Index) / sizeof(int),
```
Comment (kidder) [inline src/Parallel/ArrayIndex.hpp]: don't do integer division, (ie multiply both sides by `sizeof(int)`

File: src/DataStructures/Tensor/Tensor.hpp [hunk 2]
```diff
@@ -431,6 +431,22 @@ class Tensor<X, Symm, IndexList<Indices...>> {
   }
   //@}
 
+  /// \brief Suffix to append to the tensor name that indicates the component
+  ///
+  /// The suffix is empty for scalars and else is an underscore followed by the
```
Comment (kidder) [inline src/DataStructures/Tensor/Tensor.hpp]: 'scalars and else it '  -> `scalars, otherwise it is`

File: tests/Unit/IO/Test_VolumeData.cpp [hunk 3]
```diff
@@ -109,6 +109,16 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData", "[Unit][IO][H5]") {
...
+  {
+    INFO("Test find_observation_id");
+    std::vector<size_t> found_observation_ids{observation_values.size()};
+    std::transform(observation_values.begin(), observation_values.end(),
+                   found_observation_ids.begin(),
+                   [&volume_file](const double observation_value) noexcept {
+                     return volume_file.find_observation_id(observation_value);
+                   });
+    CHECK(found_observation_ids == observation_ids);
+  }
```
Comment (kidder) [inline tests/Unit/IO/Test_VolumeData.cpp:121]: add a test that fails

File: src/IO/H5/VolumeData.hpp [hunk 4]
```diff
@@ -105,6 +107,17 @@ class VolumeData : public h5::Object {
...
   double get_observation_value(size_t observation_id) const noexcept;
 
+  /// Find the observation ID that matches the `observation_value`
+  size_t find_observation_id(const double observation_value) const noexcept {
+    for (auto& observation_id : list_observation_ids()) {
+      if (get_observation_value(observation_id) == observation_value) {
+        return observation_id;
+      }
+    }
+    ERROR(MakeString{} << "No observation with value " << observation_value
```
Comment (kidder) [inline src/IO/H5/VolumeData.hpp]: shouldn't need `MakeString`, just `#include <ostream>`

**Review by kidder — CHANGES_REQUESTED** [review]
review with @wthrowe and @nilsdeppe

**Comment by nilsvu** [comment]
@kidder @nilsdeppe @wthrowe please check the fixups (not sure who is assigned as reviewers)

**Comment by kidder** [comment]
please squash the fixups

File: tests/Unit/IO/Test_VolumeData.cpp [hunk 5]
```diff
@@ -303,3 +313,28 @@ SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData.ComponentFormat1",
 #endif
   // clang-format off
 }
+
+// [[OutputRegex, No observation with value]]
+[[noreturn]] SPECTRE_TEST_CASE("Unit.IO.H5.VolumeData.FindNoObservationId",
+                               "[Unit][IO][H5]") {
+  // clang-format on
+  ASSERTION_TEST();
```
Comment (wthrowe) [inline tests/Unit/IO/Test_VolumeData.cpp]: You're testing an ERROR, not an ASSERT.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/IO/Test_VolumeData.cpp (hunk 5 above)
Comment (nilsvu) [reply tests/Unit/IO/Test_VolumeData.cpp]: thanks for catching this 👍 I squashed in this change

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
@kidder @wthrowe @nilsdeppe GitHub is happy but Travis is queued up quite a bit right now. You can decide if you trust the new CI enough to skip Travis or prefer to wait. Thanks for reviewing!

**Review by kidder — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@nilsleiffischer please give all reviewers an opportunity to look at the fixups...

File: src/DataStructures/Tensor/Tensor.hpp [hunk 6]
```diff
@@ -431,6 +431,22 @@ class Tensor<X, Symm, IndexList<Indices...>> {
...
+  /// The suffix is empty for scalars, otherwise it is an underscore followed by
+  /// the `Tensor::component_name` of the `tensor_index`.
+  ///
+  /// An example use case for the suffix is to label tensor components in
+  /// data files.
+  template <typename IndexType = int>
+  static std::string component_suffix(
+      const std::array<IndexType, rank()>& tensor_index =
+          std::array<IndexType, rank()>{},
+      const std::array<std::string, rank()>& axis_labels =
```
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Tensor.hpp:445]: Can you please document `axis_labels`?

**Comment by nilsvu** [comment]
@nilsdeppe sorry, I read Larry's comment as a request to squash, and also I wasn't clear who was actually assigned as reviewers to this PR, so figured you perhaps talked offline. I push your requested changes to a new PR asap.

**Comment by nilsdeppe** [comment]
No worries! Thank you :)

**Pull request merged**